accounts/manager.go

```diff
@@ -34,8 +35,34 @@ type Manager struct {
 
 	feed event.Feed // Wallet feed notifying of arrivals/departures
 
-	quit chan chan error
-	lock sync.RWMutex
+	quit      chan chan error
+	lock      sync.RWMutex
+	nonceLock accountNonceLocker // Used to lock account nonces, to prohibit nonce reuse
+}
+
+type accountNonceLocker struct {
+	mut   sync.Mutex
+	locks map[common.Address]*sync.Mutex
+}
+
+// LockNonce Locks a particular nonce for an account, to prevent another tx getting the same nonce until this lock has been released
+// This should be used when a nonce has not been provided by the user,
+// and we get a nonce from the pools. The mutex prevents the (an identical nonce) from being
+// read again during the time that the first transaction is being signed.
+func (am *Manager) LockNonce(address common.Address) {
+	if _, exist := am.nonceLock.locks[address]; !exist {
+		am.nonceLock.mut.Lock()
+		if _, exist := am.nonceLock.locks[address]; !exist {
```

Comment:
I don't like that example, for the precise reason that it keeps holding the `mut` lock needlessly, while trying to get the address-specific lock. But since you really want me to use the `defer` pattern, I'll refactor it into two methods instead, so we can avoid that.